File: python/125_ValidPalindrome.py
import re
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = re.sub('[\W_]+', '', s).lower()
        half = [len(s)//2, len(s)//2] if len(s)&1 else [len(s)//2, len(s)//2-1]
        return s[:half[0]] == s[-1:half[1]:-1] if s else True
        

File: java/125_ValidPalindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int left = 0; int right = s.length()-1;
        while (left < right) {
            if (!Character.isLetter(s.charAt(left)) && !Character.isDigit(s.charAt(left))) {
                while (left < s.length() && !Character.isLetter(s.charAt(left)) && !Character.isDigit(s.charAt(left))) ++left;
            }
            if (!Character.isLetter(s.charAt(right)) && !Character.isDigit(s.charAt(right))) {
                while (right >= 0 && !Character.isLetter(s.charAt(right)) && !Character.isDigit(s.charAt(right))) --right;
            }
            
            if (left < s.length() && right >= 0 && Character.toLowerCase(s.charAt(left++)) != Character.toLowerCase(s.charAt(right--))) return false;
        }
        return true;
    }
}
